### PR TITLE
FOGL-8980: Support multiple DNP3 outstations

### PR DESCRIPTION
**Pull request by pintomax**
FOGL-8980: Support multiple DNP3 outstations

### CODE REVIEW

**Review by MarkRiddoch — DISMISSED** [review]
Looks good to me. We need to update the documentation for the plugin

File: plugin.cpp [hunk 1]
```diff
@@ -96,6 +96,39 @@ static const char *default_config = QUOTE({
...
+					"description" : "The outstation TCP address or name ",
+					"displayName" : "TCP Address",
+					"type" : "string",
+					"default" : "127.0.0.1"
+				},
+			"port" : {
+					"description" : "The outstation TCP port",
+					"displayName" : "TCP Port",
+					"type" : "integer",
+					"options" : [ "integer","float", "boolean" ],
```
Comment (praveen-garg) [inline plugin.cpp:126]: ?

File: plugin.cpp [hunk 2]
```diff
@@ -96,6 +96,39 @@ static const char *default_config = QUOTE({
...
+		"description": "A list of DNP3 outstations to connect",
+		"type": "list",
+		"items" : "object",
+		"default": "[]",
+		"order" : "9",
+		"displayName" : "Outstations list",
+		"properties" : {
+			"linkid" : {
+					"description" : "The outstation link id",
+					"displayName" : "Link id",
```
Comment (praveen-garg) [inline plugin.cpp:110]: ID

File: plugin.cpp [hunk 3]
```diff
@@ -96,6 +96,39 @@ static const char *default_config = QUOTE({
...
 		"displayName" : "Network timeout",
 		"order" : "8"
+		},
+	"outstations": {
+		"description": "A list of DNP3 outstations to connect",
+		"type": "list",
+		"items" : "object",
+		"default": "[]",
+		"order" : "9",
+		"displayName" : "Outstations list",
```
Comment (praveen-garg) [inline plugin.cpp:106]: Remove the word _list_

**Review by praveen-garg — COMMENTED** [review]
1) As we have list of outstations so remove old config i.e. Outstation address, port, link id

2) Consider upgrade basic test and deprecate old config items

**Comment by pintomax** [comment]
> 1. As we have list of outstations so remove old config i.e. Outstation address, port, link id
> 2. Consider upgrade basic test and deprecate old config items

We have decided to keep basic configuration for the time being.

If array values are set into "outstations", those will be used

**Comment by praveen-garg** [comment]
> > 1. As we have list of outstations so remove old config i.e. Outstation address, port, link id
> > 2. Consider upgrade basic test and deprecate old config items
> 
> We have decided to keep basic configuration for the time being.
> 
> If array values are set into "outstations", those will be used

Only Outstations list value will be used, Or this existing will be considered as well?

**Comment by pintomax** [comment]
> > > 1. As we have list of outstations so remove old config i.e. Outstation address, port, link id
> > > 2. Consider upgrade basic test and deprecate old config items
> > 
> > 
> > We have decided to keep basic configuration for the time being.
> > If array values are set into "outstations", those will be used
> 
> Only Outstations list value will be used, Or this existing will be considered as well?

The existing config (one outstation) can be left there for compatibility.

If list is set and we have at least one element, the one outstation config will be ignored.

We have to update the documentation, possibly logging an info message in this case

File: dnp3.cpp [hunk 4]
```diff
@@ -172,15 +182,86 @@ bool DNP3::configure(ConfigCategory* config)
...
+		if (document.Parse(outstations.c_str()).HasParseError())
+		{
+			Logger::getLogger()->error("Error while parsing '%s' type 'list' item: %s",
+					"outstations",
+					GetParseError_En(document.GetParseError()));
+			return false;
+		}
+		if (!document.IsArray())
+		{
+			Logger::getLogger()->error("Error '%s' type 'list' item is not an arrary",
```
Comment (MarkRiddoch) [inline dnp3.cpp:200]: Typo: array -> array

File: dnp3.cpp [hunk 5]
```diff
@@ -172,15 +182,86 @@ bool DNP3::configure(ConfigCategory* config)
...
+				Logger::getLogger()->error("Error '%s' type 'list': array element is not an object",
+						"outstations");
+				return false;
+			}
+			DNP3::OutStationTCP *outstation = new DNP3::OutStationTCP();
+			for (auto& v : o.GetObject())
+                        {
+				string key = v.name.GetString();
+				string value = config->to_string(v.value);
+				Logger::getLogger()->error(">>> %s: %s", key.c_str(), value.c_str());
```
Comment (MarkRiddoch) [inline dnp3.cpp:217]: This is not a meaningful error to report - is this debug that has been left in?

**Review by MarkRiddoch — CHANGES_REQUESTED** [review]
Just a couple of minor things. Also we need to update the documentation to describe the configuration with the list versus the older, individual strings.

**Comment by pintomax** [comment]
Still need to update configuration for
- single outstation and multiple outstations
- application data logging in debug log

**Comment by MarkRiddoch** [comment]
> 1. As we have list of outstations so remove old config i.e. Outstation address, port, link id
> 2. Consider upgrade basic test and deprecate old config items

The old configuration items are present for reasons of backward compatibility. They allow for people to upgrade without breaking a current installation. The documentation will discuss how to move to the newer list if required.

**Review by MarkRiddoch — APPROVED** [review]
Lets raise a separate Jira to update the documentation

**Comment by Mohit04tomar** [comment]
I’m getting following error when connection of foglamp to the DNP3 Outstation is lost:
```
Aug 13 14:00:48 foglamp-VirtualBox FogLAMP snp #1[13227]: INFO: master_1 - Begining task: Application Poll
Aug 13 14:00:53 foglamp-VirtualBox FogLAMP snp #1[13227]: WARNING: master_1 - Timeout waiting for response
Aug 13 14:01:02 foglamp-VirtualBox FogLAMP snp #1[13227]: WARNING: master_1 - Link status request - response timeout
Aug 13 14:01:53 foglamp-VirtualBox FogLAMP snp #1[13227]: INFO: master_1 - Begining task: Application Poll
Aug 13 14:01:58 foglamp-VirtualBox FogLAMP snp #1[13227]: WARNING: master_1 - Timeout waiting for response
Aug 13 14:02:02 foglamp-VirtualBox FogLAMP snp #1[13227]: WARNING: master_1 - Link status request - response timeout
Aug 13 14:02:58 foglamp-VirtualBox FogLAMP snp #1[13227]: INFO: master_1 - Begining task: Application Poll
Aug 13 14:03:02 foglamp-VirtualBox FogLAMP snp #1[13227]: WARNING: master_1 - Link status request - response timeout
Aug 13 14:03:03 foglamp-VirtualBox FogLAMP snp #1[13227]: WARNING: master_1 - Timeout waiting for response

```
And when the network of DNP3 Outstation is up again, foglamp-south-dnp3 is still giving the same error and requires disable/enable of south service of foglamp-south-dnp3 to start ingestion of data again.

Note: It was working fine in the 2.5.0 and nightly

**Comment by pintomax** [comment]
Pkease share exact steps for the disconnection test and the simulator you’re using.This branch does not have re connection improvements yetOn 13 Aug 2024, at 11:39, Mohit04tomar ***@***.***> wrote:﻿
I’m getting following error when connection of foglamp to the DNP3 Outstation is lost:
Aug 13 14:00:48 foglamp-VirtualBox FogLAMP snp #1[13227]: INFO: master_1 - Begining task: Application Poll
Aug 13 14:00:53 foglamp-VirtualBox FogLAMP snp #1[13227]: WARNING: master_1 - Timeout waiting for response
Aug 13 14:01:02 foglamp-VirtualBox FogLAMP snp #1[13227]: WARNING: master_1 - Link status request - response timeout
Aug 13 14:01:53 foglamp-VirtualBox FogLAMP snp #1[13227]: INFO: master_1 - Begining task: Application Poll
Aug 13 14:01:58 foglamp-VirtualBox FogLAMP snp #1[13227]: WARNING: master_1 - Timeout waiting for response
Aug 13 14:02:02 foglamp-VirtualBox FogLAMP snp #1[13227]: WARNING: master_1 - Link status request - response timeout
Aug 13 14:02:58 foglamp-VirtualBox FogLAMP snp #1[13227]: INFO: master_1 - Begining task: Application Poll
Aug 13 14:03:02 foglamp-VirtualBox FogLAMP snp #1[13227]: WARNING: master_1 - Link status request - response timeout
Aug 13 14:03:03 foglamp-VirtualBox FogLAMP snp #1[13227]: WARNING: master_1 - Timeout waiting for response


And when the network of DNP3 Outstation is up again, foglamp-south-dnp3 is still giving the same error and requires disable/enable of south service of foglamp-south-dnp3 to start ingestion of data again.
Note: It was working fine in the 2.5.0 and nightly

**Comment by pintomax** [comment]
@Mohit04tomar 
Could you please add exact steps  for the disconnection tests?

All the needed details would help to reproduce this in my scenario.

It's not clear "connection is lost" and "connection is up again"

Please note in this PR there is also log for connection but no new code related to connection handling.

Also add:
- version of opendnp3 library you compiled wtih plugin with
- which DNP3 simulator you are using

**Comment by Mohit04tomar** [comment]
> @Mohit04tomar Could you please add exact steps for the disconnection tests?
> 
> All the needed details would help to reproduce this in my scenario.
> 
> It's not clear "connection is lost" and "connection is up again"
> 
> Please note in this PR there is also log for connection but no new code related to connection handling.
> 
> Also add:
> 
> * version of opendnp3 library you compiled wtih plugin with
> * which DNP3 simulator you are using

Steps to reproduce:
1. Setup and Start the DNP3 simulator (Fryr SCADA DNP3 simulator)
2. Add south service of south-dnp3 with valid config in enable mode
3. Check data ingestion (GOOD)
4. Disconnect the network of DNP3 simulator machine
5. Check syslog of FogLAMP

```Aug 13 14:00:48 foglamp-VirtualBox FogLAMP snp #1[13227]: INFO: master_1 - Begining task: Application Poll
Aug 13 14:00:53 foglamp-VirtualBox FogLAMP snp #1[13227]: WARNING: master_1 - Timeout waiting for response
Aug 13 14:01:02 foglamp-VirtualBox FogLAMP snp #1[13227]: WARNING: master_1 - Link status request - response timeout
Aug 13 14:01:53 foglamp-VirtualBox FogLAMP snp #1[13227]: INFO: master_1 - Begining task: Application Poll
Aug 13 14:01:58 foglamp-VirtualBox FogLAMP snp #1[13227]: WARNING: master_1 - Timeout waiting for response
Aug 13 14:02:02 foglamp-VirtualBox FogLAMP snp #1[13227]: WARNING: master_1 - Link status request - response timeout
Aug 13 14:02:58 foglamp-VirtualBox FogLAMP snp #1[13227]: INFO: master_1 - Begining task: Application Poll
Aug 13 14:03:02 foglamp-VirtualBox FogLAMP snp #1[13227]: WARNING: master_1 - Link status request - response timeout
Aug 13 14:03:03 foglamp-VirtualBox FogLAMP snp #1[13227]: WARNING: master_1 - Timeout waiting for response
```
6. Reconnect network of DNP3 Simulator

Actual Result: Data ingestion does not start again, and above logs are still repeating in syslog.
Expected Result: Data ingestion should start again without any issue.


**Note:** south-dnp3 plugin is compiled and packaged using the OpenDNP3 library, as specified in the requirements.sh file, version 2.3.0, for Ubuntu 20.

**Comment by pintomax** [comment]
@Mohit04tomar 

what really "Disconnect the network of DNP3 simulator machine" means?

Disconnect the ethernet cable? Something else?

Please add the details in one go

**Comment by Mohit04tomar** [comment]
@pintomax Simply turn off the Internet on the machine having simulator installed

**Comment by pintomax** [comment]
@Mohit04tomar please verify the network cable disconnection works with multiple outstations as well

**Comment by pintomax** [comment]
@Mohit04tomar I tested 2 outstations:

- outstation windows with freyrScada simulator: network unplugged
- outstation-demo unix with 'sudo iptables -A OUTPUT -p tcp --destination-port 20001 -j DROP'


cable back on windows: OK
sudo iptables -F on linux: OK

This KeepAliveFailure detection seems OK


Maybe we can decrease verbosity of changed netwowk status to debug after reconnection

**Comment by pintomax** [comment]
@Mohit04tomar latest commit addresses FOGL-9019

**Comment by praveen-garg** [comment]
@pintomax I would suggest to have [39e1a7a](https://github.com/fledge-iot/fledge-south-dnp3/pull/43/commits/39e1a7a27aafd3ae4c1825f8d47897ce02aa474b) fix against _develop_. Test, merge & close that Jira that. Then update this branch.

**Review by MarkRiddoch — APPROVED** [review]
Looks good as far as I can tell. TVA need this so we need to get this merged and tested as soon as we can, without compromising testing. Don't mind if you roll both Jirs's together or seperate them. Both Jiras are needed and need the same level of testing.

**Comment by praveen-garg** [comment]
@pintomax  I have 2 outstations and stopped one of them i.e. Outstation ID 3.
Log says:
`Aug 22 17:06:25 nerdl2-Latitude-3420-2 FogLAMP dnp3 #1[192632]: WARNING: dnp3 #1_remote_3 - End of file`

There should be **meaningful** warning with outstation details.

**Comment by pintomax** [comment]
Filtering out received log messages:

Aug 22 18:17:57 ubuntu20-lenovo Fledge DN1[95598]: WARNING: Outstation id 10 - Connection errored or closed

**Review by praveen-garg — APPROVED** [review]
Thanks @pintomax

**Pull request merged**